File: Editorial/132_Palindrome_Partitioning_II/TypeScript/4_Dynamic_Programming_Bottom_Up.ts
function minCut(s: string): number {
    let cutsDp: Array<number> = new Array<number>(s.length);
    let palindromeDp: Array<Array<boolean>> = Array.from(
        Array(s.length),
        () => new Array<boolean>(s.length),
    );
    // build the palindrome cutsDp for all susbtrings
    buildPalindromeDp(s, s.length, palindromeDp);
    for (let end = 0; end < s.length; end++) {
        let minimumCut: number = end;
        for (let start = 0; start <= end; start++) {
            if (palindromeDp[start][end]) {
                minimumCut =
                    start === 0
                        ? 0
                        : Math.min(minimumCut, cutsDp[start - 1] + 1);
            }
        }
        cutsDp[end] = minimumCut;
    }
    return cutsDp[s.length - 1];
    function buildPalindromeDp(
        s: string,
        n: number,
        palindromeDp: Array<Array<boolean>>,
    ) {
        for (let end = 0; end < s.length; end++) {
            for (let start = 0; start <= end; start++) {
                if (
                    s[start] == s[end] &&
                    (end - start <= 2 || palindromeDp[start + 1][end - 1])
                ) {
                    palindromeDp[start][end] = true;
                }
            }
        }
    }
}

File: Editorial/132_Palindrome_Partitioning_II/Python3/4_Dynamic_Programming_Bottom_Up.py
class Solution:
    def minCut(self, s: str) -> int:
        cutsDP = [0] * len(s)
        palindromeDP = [[False] * len(s) for _ in range(len(s))]
        # build the palindrome cutsDP for all substrings
        self.buildPalindromeDP(s, len(s), palindromeDP)
        for end in range(len(s)):
            minimumCut = end
            for start in range(end + 1):
                if palindromeDP[start][end]:
                    minimumCut = (
                        0
                        if start == 0
                        else min(minimumCut, cutsDP[start - 1] + 1)
                    )
            cutsDP[end] = minimumCut
        return cutsDP[len(s) - 1]

    def buildPalindromeDP(self, s, n, palindromeDP):
        for end in range(len(s)):
            for start in range(end + 1):
                if s[start] == s[end] and (
                    end - start <= 2 or palindromeDP[start + 1][end - 1]
                ):
                    palindromeDP[start][end] = True
